### PR TITLE
don't consume bookbinder inputs if the output is worse than the input

### LINKED ISSUE #3247: Book Binder reverts the enchantment level to maximal on the config
### ❗ Checklist

- [X] I am using the official english version of Slimefun and did not modify the jar.
- [X] I am using an up to date "DEV" (not "RC") version of Slimefun.
- [X] I am aware that issues related to Slimefun addons need to be reported on their bug trackers and not here.
- [X] I searched for similar open issues and could not find an existing bug report on this.

### 📍 Description

Book Binder reverts the enchantment level to maximal on the config if your enchanted book exceeds it.

https://discord.com/channels/565557184348422174/565569972697301020/886602140280324136

### 📑 Reproduction Steps

1. Set `bypass-vanilla-max-level: false` for vanilla cap or `custom-max-level: 10` and `has-custom-max-level: true`
2. Use book binder
3. Put an enchantment that exceeds vanilla limit or maximum level cap on config
4. Wait it process
5. It will revert the enchantment to maximum configured level (vanilla/config).

### 💡 Expected Behavior

Book binder should not take the items and revert the enchantment to maximum configured level (vanilla/config). It should just not work.

### 📷 Screenshots / Videos

https://youtu.be/lW0-6tL46SM

### 📜 Server Log

_No response_

### 📂 `/error-reports/` folder

_No response_

### 💻 Server Software

Paper

### 🎮 Minecraft Version

1.17.x

### ⭐ Slimefun version

```
> sf versions
[21:07:57 INFO]: This Server uses the following setup of Slimefun:
Paper git-Paper-259 (MC: 1.17.1)
Slimefun DEV - 968 (git 79c6c139)
Metrics-Module #28
Java 16

No Addons installed
> pl
[21:07:57 INFO]: Plugins (3): Slimefun, Vault, VoidGen
> ver
[21:07:58 INFO]: Checking version, please wait...
[21:08:00 INFO]: This server is running Paper version git-Paper-259 (MC: 1.17.1) (Implementing API version 1.17.1-R0.1-SNAPSHOT) (Git: 810d200)
You are running the latest version
Previous version: git-Paper-250 (MC: 1.17.1)
```

### 🧭 Other plugins

_No response_

### PR DESCRIPTION
**Pull request by iTwins**
## Description

The bookbinder used to accept any book and just reset any level higher that the configured max back to the max.
For example protection 10 + protection 10 = protection 4

## Proposed changes

If any of the books contains an enchantment level higher than the configured max: don't consume the items
If the output would be the same any of the inputs: don't consume the items

## Related Issues (if applicable)

Resolves #3247

## Checklist

- [x] I have fully tested the proposed changes and promise that they will not break everything into chaos.
- [ ] I have also tested the proposed changes in combination with various popular addons and can confirm my changes do not break them.
- [x] I have made sure that the proposed changes do not break compatibility across the supported Minecraft versions (1.16.* - 1.20.*).
- [x] I followed the existing code standards and didn't mess up the formatting.
- [x] I did my best to add documentation to any public classes or methods I added.
- [x] I have added `Nonnull` and `Nullable` annotations to my methods to indicate their behaviour for null values
- [ ] I added sufficient Unit Tests to cover my code.

### CODE REVIEW

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/electric/machines/enchanting/BookBinder.java [hunk 1]
```diff
@@ -57,6 +57,20 @@ protected MachineRecipe findNextRecipe(BlockMenu menu) {
...
+                            return null;
+                        }
+                    }
+                    for (Map.Entry<Enchantment, Integer> entry : storedTargetEnchantments.entrySet()) {
+                        int maxLevel = bypassVanillaMaxLevel.getValue() ? customMaxLevel.getValue() : entry.getKey().getMaxLevel();
+                        if (entry.getValue() > maxLevel) {
+                            return null;
+                        }
+                    }
+
```
Comment (J3fftw1) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/electric/machines/enchanting/BookBinder.java:73]: not a big fan of that we have 2 for loops with the same exact code in it

Comment (iTwins) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/electric/machines/enchanting/BookBinder.java]: Agreed, it has a helper function now. Is it better like this?

**Review by J3fftw1 — APPROVED** [review]
LGTM to just need to be tested then its ready to merge imo

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/electric/machines/enchanting/BookBinder.java [hunk 2]
```diff
@@ -97,6 +106,20 @@ private boolean isCompatible(@Nullable ItemStack item) {
         return item != null && item.getType() == Material.ENCHANTED_BOOK;
     }
 
+    private boolean hasIllegalEnchants(@Nullable Map<Enchantment, Integer> enchantments) {
+        if (enchantments == null) {
+            return false;
+        }
+
+        for (Map.Entry<Enchantment, Integer> entry : enchantments.entrySet()) {
+            int maxLevel = bypassVanillaMaxLevel.getValue() ? customMaxLevel.getValue() : entry.getKey().getMaxLevel();
```
Comment (JustAHuman-xD) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/electric/machines/enchanting/BookBinder.java:115]: Don't want to use a ternary here (See: [Code Style](https://github.com/Slimefun/Slimefun4/blob/master/CONTRIBUTING.md#7-style-preferences))

**Comment by J3fftw1** [comment]
is this PR not the same as https://github.com/Slimefun/Slimefun4/pull/3299?

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/electric/machines/enchanting/BookBinder.java (hunk 2 above)
Comment (iTwins) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/electric/machines/enchanting/BookBinder.java:115]:
```java
if (bypassVanillaMaxLevel.getValue() && entry.getValue() > customMaxLevel.getValue()) {
    return true;
} else if (entry.getValue() > entry.getKey().getMaxLevel()) {
    return true;
}
```
Would this be better?

**Comment by iTwins** [comment]
> is this PR not the same as #3299?

No it is not. i just tested the changes in that build and it did not work (prot 10 + prot 4 = prot 4). It also has some other problems like returning null when a method is annotated as ``@Nonnull`` and not covering other edge cases.

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/electric/machines/enchanting/BookBinder.java (hunk 2 above)
Comment (iTwins) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/electric/machines/enchanting/BookBinder.java:115]: or in 1 if better?
```java
if (bypassVanillaMaxLevel.getValue() && entry.getValue() > customMaxLevel.getValue() || !bypassVanillaMaxLevel.getValue() && entry.getValue() > entry.getKey().getMaxLevel()) {
    return true;
}
```

Comment (J3fftw1) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/electric/machines/enchanting/BookBinder.java:115]: second would be better

Comment (iTwins) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/electric/machines/enchanting/BookBinder.java]: Fixed

**Comment by variananora** [comment]
Tested in 1.20.1 and it works flawlessly.

**Review by J3fftw1 — APPROVED** [review]

**Review by Sefiraat — APPROVED** [review]

**Pull request merged**